File: benchmarks/scala-stm/scala-stm-library/lib/sb7-java-v1.2/src/stmbench7/impl/core/ModuleImpl.java
package stmbench7.impl.core;

import stmbench7.core.ComplexAssembly;
import stmbench7.core.Manual;
import stmbench7.core.Module;


/**
 * STMBench7 benchmark Module (see the specification).
 * Default implementation.
 */
public class ModuleImpl extends DesignObjImpl implements Module {

    private final Manual man;
    private ComplexAssembly designRoot;

    public ModuleImpl(int id, String type, int buildDate, Manual man) {
    	super(id, type, buildDate);
    	this.man = man;
    	man.setModule(this);
    }

    public ModuleImpl(ModuleImpl source) {
    	super(source);
    	this.man = source.man;
    	this.designRoot = source.designRoot;
    }
    
    public void setDesignRoot(ComplexAssembly designRoot) {
    	this.designRoot = designRoot;
    }
    
    public ComplexAssembly getDesignRoot() {
    	return designRoot;
    }

    public Manual getManual() {
    	return man;
    }
    
	@Override
	public boolean equals(Object obj) {
		if(! (obj instanceof Module)) return false;
		return super.equals(obj);
	}
}

File: benchmarks/scala-stm/src/main/scala/stmbench7/scalastm/ModuleImpl.scala
/* scala-stm - (c) 2009-2011, Stanford University, PPL */

package stmbench7.scalastm

import scala.concurrent.stm.Ref

import stmbench7.core.ComplexAssembly
import stmbench7.core.Manual
import stmbench7.core.Module

class ModuleImpl(id: Int, typ: String, buildDate: Int, man: Manual)
  extends DesignObjImpl(id, typ, buildDate)
  with Module {

  private val designRoot = Ref(null: ComplexAssembly).single

  man.setModule(this)

  override def setDesignRoot(v: ComplexAssembly): Unit = { designRoot() = v }
  override def getDesignRoot: ComplexAssembly = designRoot()
  override def getManual: Manual = man
}
